File: urbanairship-core/src/main/java/com/urbanairship/permission/Permission.java
/* Copyright Airship and Contributors */

package com.urbanairship.permission;

import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonSerializable;
import com.urbanairship.json.JsonValue;

import java.util.Locale;

import androidx.annotation.NonNull;

/**
 * Device permissions.
 */
public enum Permission implements JsonSerializable {
    // Display notifications
    DISPLAY_NOTIFICATIONS("display_notifications"),

    // Access location
    LOCATION("location");

    @NonNull
    private final String value;

    Permission(@NonNull String value) {
        this.value = value;
    }

    @NonNull
    public String getValue() {
        return value;
    }

    @NonNull
    public static Permission fromJson(@NonNull JsonValue value) throws JsonException {
        String valueString = value.optString();
        for (Permission type : Permission.values()) {
            if (type.value.equalsIgnoreCase(valueString)) {
                return type;
            }
        }
        throw new JsonException("Invalid permission: " + value);
    }

    @NonNull
    @Override
    public String toString() {
        return name().toLowerCase(Locale.ROOT);
    }

    @NonNull
    @Override
    public JsonValue toJsonValue() {
        return JsonValue.wrapOpt(value);
    }
}

File: urbanairship-core/src/main/java/com/urbanairship/permission/Permission.kt
/* Copyright Airship and Contributors */
package com.urbanairship.permission

import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonSerializable
import com.urbanairship.json.JsonValue

/**
 * Device permissions.
 *
 * Only [DISPLAY_NOTIFICATIONS] has a built-in delegate. Requesting any other permission
 * requires the app to register a [PermissionDelegate] with [PermissionsManager], otherwise
 * the request resolves to [PermissionStatus.NOT_DETERMINED] and nothing is shown.
 */
public enum class Permission(public val value: String) : JsonSerializable {

    // Display notifications
    DISPLAY_NOTIFICATIONS("display_notifications"),

    // Access location
    LOCATION("location"),

    // App Tracking Transparency. iOS only; Android has no equivalent OS prompt, so this
    // resolves to NOT_DETERMINED unless the app maps it to a consent flow of its own.
    APP_TRACKING_TRANSPARENCY("app_tracking_transparency"),

    // Camera
    CAMERA("camera"),

    // Microphone
    MICROPHONE("microphone"),

    // Bluetooth
    BLUETOOTH("bluetooth"),

    // Photo library
    PHOTO_LIBRARY("photo_library"),

    // Contacts
    CONTACTS("contacts");

    override fun toString(): String = name.lowercase()

    override fun toJsonValue(): JsonValue = JsonValue.wrapOpt(value)

    public companion object {

        @Throws(JsonException::class)
        public fun fromJson(value: JsonValue): Permission {
            val content = value.requireString().lowercase()
            return entries.firstOrNull { it.value == content }
                ?: throw JsonException("Invalid permission: $value")
        }
    }
}
